src/snappy_device_agents/devices/maas2/maas_storage.py

```diff
@@ -0,0 +1,560 @@
+# Copyright (C) 2023 Canonical
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Ubuntu MaaS 2.x storage provisioning code."""
+
+import logging
+import subprocess
+import collections
+import json
+import math
+
+
+logger = logging.getLogger()
+
+
+class MaasStorageError(Exception):
+    pass
+
+
+class MaasStorage:
+    """Maas device agent storage module."""
+
+    def __init__(self, maas_user, node_id):
+        self.maas_user = maas_user
+        self.node_id = node_id
+        self.device_list = None
+        self.init_data = None
+        self.node_info = self._node_read()
+        self.block_ids = {}
+        self.partition_sizes = {}
+
+    def _logger_debug(self, message):
+        logger.debug("MAAS: {}".format(message))
+
+    def _logger_info(self, message):
+        logger.info("MAAS: {}".format(message))
+
+    def _node_read(self):
+        """Read node block-devices.
+
+        :return: the node's block device information
+        """
+        cmd = ["maas", self.maas_user, "block-devices", "read", self.node_id]
+        return self.call_cmd(cmd, output_json=True)
+
+    @staticmethod
+    def call_cmd(cmd, output_json=False):
+        """Run a command and return the output.
+
+        :param cmd: command to run
+        :param output_json: output the result as JSON
+        :return: subprocess stdout
+        :raises MaasStorageError: on subprocess non-zero return code
+        """
+        try:
+            proc = subprocess.run(
+                cmd,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+                check=False,
+            )
+        except subprocess.CalledProcessError:
+            raise MaasStorageError(proc.stdout.decode())
```

Comment:
This won't work. subprocess.run() is never going to raise this because you passed check=False. However, don't just take that out either. If you do, then proc will be undefined and you'll get a different exception here other than what you intend and it will be confusing.

You can either call proc.check_returncode() under the try block instead. Or you can just check if proc.returncode != 0, and raise the error, which I think is probably nicer.

Reply:
This should never have been refactored in the first place, this is an artifact from a test method that was not running check=False. This was an oversight on my part, thanks for catching that.